File: python_implementation/liquidity_stress_test.py
"""
Lloyd's Liquidity Stress Test - Python Implementation for Power BI
Processes Lloyd's syndicate data to generate RRA (Reserve Return Annual) form outputs
"""

import pandas as pd
import numpy as np
from datetime import datetime
import json
import os

class LiquidityStressTest:
    """
    Main class for processing Lloyd's Liquidity Stress Test data
    Compatible with Power BI Python visual
    """

    def __init__(self, data_path='../data'):
        """
        Initialize with path to data directory

        Args:
            data_path (str): Path to directory containing syndicate data
        """
        self.data_path = data_path
        self.syndicates = []
        self.metadata = None
        self.assets_data = None
        self.cashflow_data = None
        self.stress_data = None

    def load_data(self, syndicate_number=None):
        """
        Load data for analysis

        Args:
            syndicate_number (int, optional): Specific syndicate to load.
                                             If None, loads all syndicates.

        Returns:
            dict: Loaded data dictionary
        """
        if syndicate_number:
            # Load individual syndicate
            syndicate_path = os.path.join(self.data_path, f'syndicate_{syndicate_number}')

            with open(os.path.join(syndicate_path, 'metadata.json'), 'r') as f:
                metadata = json.load(f)

            assets = pd.read_csv(os.path.join(syndicate_path, 'assets_liquidity.csv'))
            cashflow = pd.read_csv(os.path.join(syndicate_path, 'cashflow.csv'))
            stress = pd.read_csv(os.path.join(syndicate_path, 'stress_scenario.csv'))

            return {
                'metadata': metadata,
                'assets': assets,
                'cashflow': cashflow,
                'stress': stress
            }
        else:
            # Load all syndicates combined
            self.metadata = pd.read_csv(os.path.join(self.data_path, 'all_syndicates_metadata.csv'))
            self.assets_data = pd.read_csv(os.path.join(self.data_path, 'all_syndicates_assets.csv'))
            self.cashflow_data = pd.read_csv(os.path.join(self.data_path, 'all_syndicates_cashflow.csv'))
            self.stress_data = pd.read_csv(os.path.join(self.data_path, 'all_syndicates_stress.csv'))

            return {
                'metadata': self.metadata,
                'assets': self.assets_data,
                'cashflow': self.cashflow_data,
                'stress': self.stress_data
            }

    def calculate_capital_position_table(self, syndicate_number=None):
        """
        Generate Capital Position summary table for RRA forms

        Args:
            syndicate_number (int, optional): Filter for specific syndicate

        Returns:
            pd.DataFrame: Capital position table
        """
        if self.metadata is None:
            self.load_data()

        df = self.metadata.copy()

        if syndicate_number:
            df = df[df['syndicate_number'] == syndicate_number]

        # Select relevant columns
        capital_position = df[[
            'syndicate_number',
            'syndicate_name',
            'managing_agent',
            'qma_date',
            'syndicate_fal',
            'syndicate_fis',
            'syndicate_uscr',
            'syndicate_ueca'
        ]].copy()

        # Calculate key ratios
        capital_position['solvency_ratio'] = (
            capital_position['syndicate_fal'] / capital_position['syndicate_uscr']
        ).round(2)

        capital_position['eca_coverage'] = (
            capital_position['syndicate_fal'] / capital_position['syndicate_ueca']
        ).round(2)

        # Format currency columns
        currency_cols = ['syndicate_fal', 'syndicate_fis', 'syndicate_uscr', 'syndicate_ueca']
        for col in currency_cols:
            capital_position[f'{col}_formatted'] = capital_position[col].apply(
                lambda x: f'£{x:,.0f}'
            )

        return capital_position

    def calculate_liquidity_breakdown_table(self, syndicate_number=None, date=None):
        """
        Generate Asset & Liquidity Breakdown table for RRA forms

        Args:
            syndicate_number (int, optional): Filter for specific syndicate
            date (str, optional): Filter for specific date (YYYY-MM-DD)

        Returns:
            pd.DataFrame: Liquidity breakdown table
        """
        if self.assets_data is None:
            self.load_data()

        df = self.assets_data.copy()

        if syndicate_number:
            df = df[df['syndicate_number'] == syndicate_number]

        if date:
            df = df[df['date'] == date]

        # Calculate totals and percentages
        df['total_assets'] = (
            df['restricted_assets_total'] +
            df['illiquid_assets_total'] +
            df['liquid_assets_total']
        )

        df['restricted_pct'] = (df['restricted_assets_total'] / df['total_assets'] * 100).round(2)
        df['illiquid_pct'] = (df['illiquid_assets_total'] / df['total_assets'] * 100).round(2)
        df['liquid_pct'] = (df['liquid_assets_total'] / df['total_assets'] * 100).round(2)

        return df

    def calculate_cashflow_summary_table(self, syndicate_number=None):
        """
        Generate Cashflow Summary table for RRA forms

        Args:
            syndicate_number (int, optional): Filter for specific syndicate

        Returns:
            pd.DataFrame: Cashflow summary table
        """
        if self.cashflow_data is None:
            self.load_data()

        df = self.cashflow_data.copy()

        if syndicate_number:
            df = df[df['syndicate_number'] == syndicate_number]

        # Add cumulative columns
        for syndicate in df['syndicate_number'].unique():
            mask = df['syndicate_number'] == syndicate

            df.loc[mask, 'cumulative_premium_income'] = df.loc[mask, 'premium_income'].cumsum()
            df.loc[mask, 'cumulative_claims_paid'] = df.loc[mask, 'claims_paid'].cumsum()
            df.loc[mask, 'cumulative_total_movements'] = df.loc[mask, 'total_movements'].cumsum()

        return df

    def calculate_stress_impact_table(self, syndicate_number=None):
        """
        Generate Stress Test Impact table for RRA forms

        Args:
            syndicate_number (int, optional): Filter for specific syndicate

        Returns:
            pd.DataFrame: Stress impact table with baseline vs stressed scenarios
        """
        if self.cashflow_data is None or self.stress_data is None:
            self.load_data()

        # Merge baseline cashflow with stress scenario
        baseline = self.cashflow_data.copy()
        stress = self.stress_data.copy()

        merged = baseline.merge(
            stress,
            on=['syndicate_number', 'date'],
            how='left'
        )

        if syndicate_number:
            merged = merged[merged['syndicate_number'] == syndicate_number]

        # Calculate stressed cashflow positions
        merged['stressed_closing_funds'] = (
            merged['closing_free_funds'] + merged['stress_scenario_impact']
        )

        merged['liquidity_gap'] = merged['closing_free_funds'] - merged['stressed_closing_funds']

        # Calculate minimum liquidity position
        for syndicate in merged['syndicate_number'].unique():
            mask = merged['syndicate_number'] == syndicate
            merged.loc[mask, 'min_liquidity_baseline'] = merged.loc[mask, 'closing_free_funds'].min()
            merged.loc[mask, 'min_liquidity_stressed'] = merged.loc[mask, 'stressed_closing_funds'].min()

        return merged

    def calculate_rra_output_tables(self, syndicate_number=None):
        """
        Generate all RRA output tables

        Args:
            syndicate_number (int, optional): Filter for specific syndicate

        Returns:
            dict: Dictionary containing all output tables
        """
        return {
            'capital_position': self.calculate_capital_position_table(syndicate_number),
            'liquidity_breakdown': self.calculate_liquidity_breakdown_table(syndicate_number),
            'cashflow_summary': self.calculate_cashflow_summary_table(syndicate_number),
            'stress_impact': self.calculate_stress_impact_table(syndicate_number)
        }

    def create_dashboard_summary(self, syndicate_number=None):
        """
        Create executive dashboard summary with key metrics

        Args:
            syndicate_number (int, optional): Filter for specific syndicate

        Returns:
            pd.DataFrame: Dashboard summary metrics
        """
        if self.metadata is None:
            self.load_data()

        capital = self.calculate_capital_position_table(syndicate_number)
        stress = self.calculate_stress_impact_table(syndicate_number)

        summary_data = []

        for _, synd_capital in capital.iterrows():
            synd_num = synd_capital['syndicate_number']

            synd_stress = stress[stress['syndicate_number'] == synd_num]

            # Get stress scenario metadata
            synd_meta = self.metadata[self.metadata['syndicate_number'] == synd_num].iloc[0]

            summary = {
                'syndicate_number': synd_num,
                'syndicate_name': synd_capital['syndicate_name'],
                'managing_agent': synd_capital['managing_agent'],
                'total_fal': synd_capital['syndicate_fal'],
                'solvency_ratio': synd_capital['solvency_ratio'],
                'baseline_min_liquidity': synd_stress['min_liquidity_baseline'].iloc[0],
                'stressed_min_liquidity': synd_stress['min_liquidity_stressed'].iloc[0],
                'max_liquidity_gap': synd_stress['liquidity_gap'].max(),
                'stress_scenario': synd_meta['scenario_type'],
                'gross_loss_estimate': synd_meta['gross_loss_estimate'],
                'net_loss_estimate': synd_meta['net_loss_estimate'],
                'us_funding_requirement': synd_meta['us_funding_requirement'],
                'stress_test_pass': synd_stress['stressed_closing_funds'].min() > 0
            }

            summary_data.append(summary)

        return pd.DataFrame(summary_data)

    def export_to_excel(self, output_path, syndicate_number=None):
        """
        Export all tables to Excel workbook for easy viewing

        Args:
            output_path (str): Path for output Excel file
            syndicate_number (int, optional): Filter for specific syndicate
        """
        tables = self.calculate_rra_output_tables(syndicate_number)
        summary = self.create_dashboard_summary(syndicate_number)

        with pd.ExcelWriter(output_path, engine='openpyxl') as writer:
            summary.to_excel(writer, sheet_name='Dashboard Summary', index=False)
            tables['capital_position'].to_excel(writer, sheet_name='Capital Position', index=False)
            tables['liquidity_breakdown'].to_excel(writer, sheet_name='Liquidity Breakdown', index=False)
            tables['cashflow_summary'].to_excel(writer, sheet_name='Cashflow Summary', index=False)
            tables['stress_impact'].to_excel(writer, sheet_name='Stress Impact Analysis', index=False)

        print(f"✓ Exported to {output_path}")


# Power BI Integration Functions
# These functions can be called directly from Power BI Python visuals

def get_capital_position(dataset):
    """
    Power BI wrapper for capital position table

    Args:
        dataset (pd.DataFrame): Input dataset from Power BI

    Returns:
        pd.DataFrame: Processed capital position table
    """
    lst = LiquidityStressTest()
    lst.metadata = dataset
    return lst.calculate_capital_position_table()


def get_liquidity_breakdown(dataset):
    """
    Power BI wrapper for liquidity breakdown table

    Args:
        dataset (pd.DataFrame): Input dataset from Power BI

    Returns:
        pd.DataFrame: Processed liquidity breakdown table
    """
    lst = LiquidityStressTest()
    lst.assets_data = dataset
    return lst.calculate_liquidity_breakdown_table()


def get_stress_impact(cashflow_dataset, stress_dataset):
    """
    Power BI wrapper for stress impact analysis

    Args:
        cashflow_dataset (pd.DataFrame): Cashflow data from Power BI
        stress_dataset (pd.DataFrame): Stress scenario data from Power BI

    Returns:
        pd.DataFrame: Processed stress impact table
    """
    lst = LiquidityStressTest()
    lst.cashflow_data = cashflow_dataset
    lst.stress_data = stress_dataset
    return lst.calculate_stress_impact_table()


# Example usage and testing
if __name__ == "__main__":
    print("Lloyd's Liquidity Stress Test - Python Implementation")
    print("=" * 80)

    # Initialize
    lst = LiquidityStressTest('data')

    # Load all data
    print("\nLoading data...")
    lst.load_data()
    print(f"✓ Loaded data for {len(lst.metadata)} syndicates")

    # Generate dashboard summary
    print("\nGenerating dashboard summary...")
    summary = lst.create_dashboard_summary()
    print(summary.to_string())

    # Export all tables for syndicate 2001
    print("\n" + "=" * 80)
    print("Exporting detailed analysis for Syndicate 2001...")
    lst.export_to_excel('python_implementation/syndicate_2001_analysis.xlsx', syndicate_number=2001)

    # Export combined analysis for all syndicates
    print("\nExporting combined analysis for all syndicates...")
    lst.export_to_excel('python_implementation/all_syndicates_analysis.xlsx')

    print("\n" + "=" * 80)
    print("Processing complete!")

File: archive/v1.2.0/r_implementation/liquidity_stress_test.R
# Lloyd's Liquidity Stress Test - R Implementation for Power BI
# Processes Lloyd's syndicate data to generate RRA (Reserve Return Annual) form outputs

library(dplyr)
library(tidyr)
library(jsonlite)
library(readr)
library(openxlsx)

#' LiquidityStressTest R6 Class
#'
#' Main class for processing Lloyd's Liquidity Stress Test data
#' Compatible with Power BI R visual
#'
#' @export
LiquidityStressTest <- R6::R6Class(
  "LiquidityStressTest",

  public = list(
    data_path = NULL,
    metadata = NULL,
    assets_data = NULL,
    cashflow_data = NULL,
    stress_data = NULL,

    #' Initialize the LiquidityStressTest object
    #'
    #' @param data_path Path to directory containing syndicate data
    initialize = function(data_path = "../data") {
      self$data_path <- data_path
    },

    #' Load data for analysis
    #'
    #' @param syndicate_number Specific syndicate to load (optional)
    #' @return List of loaded data
    load_data = function(syndicate_number = NULL) {
      if (!is.null(syndicate_number)) {
        # Load individual syndicate
        syndicate_path <- file.path(self$data_path, paste0("syndicate_", syndicate_number))

        metadata <- fromJSON(file.path(syndicate_path, "metadata.json"))
        assets <- read_csv(file.path(syndicate_path, "assets_liquidity.csv"), show_col_types = FALSE)
        cashflow <- read_csv(file.path(syndicate_path, "cashflow.csv"), show_col_types = FALSE)
        stress <- read_csv(file.path(syndicate_path, "stress_scenario.csv"), show_col_types = FALSE)

        return(list(
          metadata = metadata,
          assets = assets,
          cashflow = cashflow,
          stress = stress
        ))
      } else {
        # Load all syndicates combined
        self$metadata <- read_csv(file.path(self$data_path, "all_syndicates_metadata.csv"), show_col_types = FALSE)
        self$assets_data <- read_csv(file.path(self$data_path, "all_syndicates_assets.csv"), show_col_types = FALSE)
        self$cashflow_data <- read_csv(file.path(self$data_path, "all_syndicates_cashflow.csv"), show_col_types = FALSE)
        self$stress_data <- read_csv(file.path(self$data_path, "all_syndicates_stress.csv"), show_col_types = FALSE)

        return(list(
          metadata = self$metadata,
          assets = self$assets_data,
          cashflow = self$cashflow_data,
          stress = self$stress_data
        ))
      }
    },

    #' Calculate Capital Position Table
    #'
    #' Generate Capital Position summary table for RRA forms
    #'
    #' @param syndicate_number Filter for specific syndicate (optional)
    #' @return Data frame with capital position
    calculate_capital_position_table = function(syndicate_number = NULL) {
      if (is.null(self$metadata)) {
        self$load_data()
      }

      df <- self$metadata

      if (!is.null(syndicate_number)) {
        df <- df %>% filter(syndicate_number == !!syndicate_number)
      }

      # Select relevant columns and calculate ratios
      capital_position <- df %>%
        select(
          syndicate_number,
          syndicate_name,
          managing_agent,
          qma_date,
          syndicate_fal,
          syndicate_fis,
          syndicate_uscr,
          syndicate_ueca
        ) %>%
        mutate(
          solvency_ratio = round(syndicate_fal / syndicate_uscr, 2),
          eca_coverage = round(syndicate_fal / syndicate_ueca, 2),
          syndicate_fal_formatted = paste0("£", format(syndicate_fal, big.mark = ",", scientific = FALSE)),
          syndicate_fis_formatted = paste0("£", format(syndicate_fis, big.mark = ",", scientific = FALSE)),
          syndicate_uscr_formatted = paste0("£", format(syndicate_uscr, big.mark = ",", scientific = FALSE)),
          syndicate_ueca_formatted = paste0("£", format(syndicate_ueca, big.mark = ",", scientific = FALSE))
        )

      return(capital_position)
    },

    #' Calculate Liquidity Breakdown Table
    #'
    #' Generate Asset & Liquidity Breakdown table for RRA forms
    #'
    #' @param syndicate_number Filter for specific syndicate (optional)
    #' @param date Filter for specific date (optional)
    #' @return Data frame with liquidity breakdown
    calculate_liquidity_breakdown_table = function(syndicate_number = NULL, date = NULL) {
      if (is.null(self$assets_data)) {
        self$load_data()
      }

      df <- self$assets_data

      if (!is.null(syndicate_number)) {
        df <- df %>% filter(syndicate_number == !!syndicate_number)
      }

      if (!is.null(date)) {
        df <- df %>% filter(date == !!date)
      }

      # Calculate totals and percentages
      df <- df %>%
        mutate(
          total_assets = restricted_assets_total + illiquid_assets_total + liquid_assets_total,
          restricted_pct = round(restricted_assets_total / total_assets * 100, 2),
          illiquid_pct = round(illiquid_assets_total / total_assets * 100, 2),
          liquid_pct = round(liquid_assets_total / total_assets * 100, 2)
        )

      return(df)
    },

    #' Calculate Cashflow Summary Table
    #'
    #' Generate Cashflow Summary table for RRA forms
    #'
    #' @param syndicate_number Filter for specific syndicate (optional)
    #' @return Data frame with cashflow summary
    calculate_cashflow_summary_table = function(syndicate_number = NULL) {
      if (is.null(self$cashflow_data)) {
        self$load_data()
      }

      df <- self$cashflow_data

      if (!is.null(syndicate_number)) {
        df <- df %>% filter(syndicate_number == !!syndicate_number)
      }

      # Add cumulative columns
      df <- df %>%
        group_by(syndicate_number) %>%
        mutate(
          cumulative_premium_income = cumsum(premium_income),
          cumulative_claims_paid = cumsum(claims_paid),
          cumulative_total_movements = cumsum(total_movements)
        ) %>%
        ungroup()

      return(df)
    },

    #' Calculate Stress Impact Table
    #'
    #' Generate Stress Test Impact table for RRA forms
    #'
    #' @param syndicate_number Filter for specific syndicate (optional)
    #' @return Data frame with stress impact analysis
    calculate_stress_impact_table = function(syndicate_number = NULL) {
      if (is.null(self$cashflow_data) || is.null(self$stress_data)) {
        self$load_data()
      }

      # Merge baseline cashflow with stress scenario
      merged <- self$cashflow_data %>%
        left_join(
          self$stress_data,
          by = c("syndicate_number", "date")
        )

      if (!is.null(syndicate_number)) {
        merged <- merged %>% filter(syndicate_number == !!syndicate_number)
      }

      # Calculate stressed cashflow positions
      merged <- merged %>%
        mutate(
          stressed_closing_funds = closing_free_funds + stress_scenario_impact,
          liquidity_gap = closing_free_funds - stressed_closing_funds
        ) %>%
        group_by(syndicate_number) %>%
        mutate(
          min_liquidity_baseline = min(closing_free_funds),
          min_liquidity_stressed = min(stressed_closing_funds)
        ) %>%
        ungroup()

      return(merged)
    },

    #' Calculate all RRA Output Tables
    #'
    #' Generate all RRA output tables
    #'
    #' @param syndicate_number Filter for specific syndicate (optional)
    #' @return List of all output tables
    calculate_rra_output_tables = function(syndicate_number = NULL) {
      list(
        capital_position = self$calculate_capital_position_table(syndicate_number),
        liquidity_breakdown = self$calculate_liquidity_breakdown_table(syndicate_number),
        cashflow_summary = self$calculate_cashflow_summary_table(syndicate_number),
        stress_impact = self$calculate_stress_impact_table(syndicate_number)
      )
    },

    #' Create Dashboard Summary
    #'
    #' Create executive dashboard summary with key metrics
    #'
    #' @param syndicate_number Filter for specific syndicate (optional)
    #' @return Data frame with dashboard summary
    create_dashboard_summary = function(syndicate_number = NULL) {
      if (is.null(self$metadata)) {
        self$load_data()
      }

      capital <- self$calculate_capital_position_table(syndicate_number)
      stress <- self$calculate_stress_impact_table(syndicate_number)

      summary_data <- capital %>%
        rowwise() %>%
        mutate(
          synd_stress = list(stress %>% filter(syndicate_number == .data$syndicate_number)),
          synd_meta = list(self$metadata %>% filter(syndicate_number == .data$syndicate_number) %>% slice(1))
        ) %>%
        mutate(
          baseline_min_liquidity = min(synd_stress[[1]]$closing_free_funds),
          stressed_min_liquidity = min(synd_stress[[1]]$stressed_closing_funds),
          max_liquidity_gap = max(synd_stress[[1]]$liquidity_gap),
          stress_scenario = synd_meta[[1]]$scenario_type,
          gross_loss_estimate = synd_meta[[1]]$gross_loss_estimate,
          net_loss_estimate = synd_meta[[1]]$net_loss_estimate,
          us_funding_requirement = synd_meta[[1]]$us_funding_requirement,
          stress_test_pass = stressed_min_liquidity > 0
        ) %>%
        select(
          syndicate_number,
          syndicate_name,
          managing_agent,
          total_fal = syndicate_fal,
          solvency_ratio,
          baseline_min_liquidity,
          stressed_min_liquidity,
          max_liquidity_gap,
          stress_scenario,
          gross_loss_estimate,
          net_loss_estimate,
          us_funding_requirement,
          stress_test_pass
        ) %>%
        ungroup()

      return(summary_data)
    },

    #' Export to Excel
    #'
    #' Export all tables to Excel workbook for easy viewing
    #'
    #' @param output_path Path for output Excel file
    #' @param syndicate_number Filter for specific syndicate (optional)
    export_to_excel = function(output_path, syndicate_number = NULL) {
      tables <- self$calculate_rra_output_tables(syndicate_number)
      summary <- self$create_dashboard_summary(syndicate_number)

      wb <- createWorkbook()

      addWorksheet(wb, "Dashboard Summary")
      writeData(wb, "Dashboard Summary", summary)

      addWorksheet(wb, "Capital Position")
      writeData(wb, "Capital Position", tables$capital_position)

      addWorksheet(wb, "Liquidity Breakdown")
      writeData(wb, "Liquidity Breakdown", tables$liquidity_breakdown)

      addWorksheet(wb, "Cashflow Summary")
      writeData(wb, "Cashflow Summary", tables$cashflow_summary)

      addWorksheet(wb, "Stress Impact Analysis")
      writeData(wb, "Stress Impact Analysis", tables$stress_impact)

      saveWorkbook(wb, output_path, overwrite = TRUE)
      cat(paste0("✓ Exported to ", output_path, "\n"))
    }
  )
)

# Power BI Integration Functions
# These functions can be called directly from Power BI R visuals

#' Get Capital Position Table
#'
#' Power BI wrapper for capital position table
#'
#' @param dataset Input dataset from Power BI
#' @return Processed capital position table
#' @export
get_capital_position <- function(dataset) {
  lst <- LiquidityStressTest$new()
  lst$metadata <- dataset
  return(lst$calculate_capital_position_table())
}

#' Get Liquidity Breakdown Table
#'
#' Power BI wrapper for liquidity breakdown table
#'
#' @param dataset Input dataset from Power BI
#' @return Processed liquidity breakdown table
#' @export
get_liquidity_breakdown <- function(dataset) {
  lst <- LiquidityStressTest$new()
  lst$assets_data <- dataset
  return(lst$calculate_liquidity_breakdown_table())
}

#' Get Stress Impact Analysis
#'
#' Power BI wrapper for stress impact analysis
#'
#' @param cashflow_dataset Cashflow data from Power BI
#' @param stress_dataset Stress scenario data from Power BI
#' @return Processed stress impact table
#' @export
get_stress_impact <- function(cashflow_dataset, stress_dataset) {
  lst <- LiquidityStressTest$new()
  lst$cashflow_data <- cashflow_dataset
  lst$stress_data <- stress_dataset
  return(lst$calculate_stress_impact_table())
}

# Example usage and testing
if (interactive() || !exists("test_mode")) {
  cat("Lloyd's Liquidity Stress Test - R Implementation\n")
  cat(strrep("=", 80), "\n")

  # Initialize
  lst <- LiquidityStressTest$new("data")

  # Load all data
  cat("\nLoading data...\n")
  lst$load_data()
  cat(paste0("✓ Loaded data for ", nrow(lst$metadata), " syndicates\n"))

  # Generate dashboard summary
  cat("\nGenerating dashboard summary...\n")
  summary <- lst$create_dashboard_summary()
  print(summary)

  # Export all tables for syndicate 2001
  cat("\n", strrep("=", 80), "\n", sep = "")
  cat("Exporting detailed analysis for Syndicate 2001...\n")
  lst$export_to_excel("r_implementation/syndicate_2001_analysis.xlsx", syndicate_number = 2001)

  # Export combined analysis for all syndicates
  cat("\nExporting combined analysis for all syndicates...\n")
  lst$export_to_excel("r_implementation/all_syndicates_analysis.xlsx")

  cat("\n", strrep("=", 80), "\n", sep = "")
  cat("Processing complete!\n")
}
